File: python/two-bucket/two_bucket.py
from math import gcd


def measure(bucket_one, bucket_two, goal, start_bucket):
    if goal % gcd(bucket_one, bucket_two) != 0 or goal > max(bucket_one, bucket_two):
        raise ValueError("Solution is impossible")

    steps = 1
    found = None
    seen = {(0, 0), (bucket_one, 0), (0, bucket_two)}
    status = set([(bucket_one, 0), ] if start_bucket == "one" else [(0, bucket_two), ])

    # have we already found a solution?
    for (v1, v2) in status:
        if v1 == goal or v2 == goal:
            found = (v1, v2)

    while found is None:
        prev_status = status
        status = set()
        steps += 1
        for (vol1, vol2) in prev_status:
            # empty/fill one bucket:
            next_steps = {(vol1, 0), (0, vol2), (vol1, bucket_two), (bucket_one, vol2)}

            # transfers:
            transfer_1to2 = min(vol1, bucket_two - vol2)
            next_steps.add((vol1 - transfer_1to2, vol2 + transfer_1to2))

            transfer_2to1 = min(vol2, bucket_one - vol1)
            next_steps.add((vol1 + transfer_2to1, vol2 - transfer_2to1))

            for n in (next_steps - seen):
                status.add(n)
                seen.add(n)

        # have we found a solution?
        for (v1, v2) in status:
            if v1 == goal or v2 == goal:
                found = (v1, v2)
                break

    if found[0] == goal:
        return steps, "one", found[1]
    else:
        return steps, "two", found[0]

File: csharp/two-bucket/TwoBucket.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Bucket
{
    One,
    Two
}

public class TwoBucketResult
{
    public int Moves { get; init; }
    public Bucket GoalBucket { get; init; }
    public int OtherBucket { get; init; }
}

public class TwoBucket(int bucketOne, int bucketTwo, Bucket startBucket)
{
     private int Bucket1 { get; } = bucketOne;
    private int Bucket2 { get; } = bucketTwo;
    private Bucket StartBucket { get; } = startBucket;


    public TwoBucketResult Measure(int goal)
    {
        if (goal % Gcd((uint) Bucket1, (uint) Bucket2) != 0 || goal > Math.Max(Bucket1, Bucket2))
            throw new ArgumentException("Solution is impossible");

        HashSet<(int vol1, int vol2)> status =
        [
            StartBucket == Bucket.One ? (Bucket1, 0) : (0, Bucket2) 
        ];
        (int vol1, int vol2) found = (0, 0);
        bool isFound = false;
        foreach (var s in status)
        {
            if (s.vol1 == goal || s.vol2 == goal)
            {
                found = s;
                isFound = true;
                break;
            }
        }
    
        int steps = 1;
        HashSet<(int vol1, int vol2)> seen = [(0, 0), (Bucket1, 0), (0, Bucket2)];
        
        while (!isFound)
        {
            HashSet<(int vol1, int vol2)> prevStatus = status;
            status = [];
            steps++;

            foreach ((int vol1, int vol2) s in prevStatus)
            {
                // empty/fill one bucket:
                HashSet<(int vol1, int vol2)> nextSteps = [
                    (s.vol1, 0 ),
                    (0, s.vol2 ),
                    (s.vol1, Bucket2),
                    (Bucket1, s.vol2)
                ];
                
                // transfers:
                int transfer1To2 = Math.Min(s.vol1, Bucket2 - s.vol2);
                nextSteps.Add( (s.vol1 - transfer1To2, s.vol2 + transfer1To2) );
                
                int transfer2To1 = Math.Min(s.vol2, Bucket1 - s.vol1);
                nextSteps.Add( (s.vol1 + transfer2To1, s.vol2 - transfer2To1) );
                
                foreach (var n in nextSteps.Except(seen))
                {
                    seen.Add(n);
                    status.Add(n);
                }
            }
            
            foreach (var s in status.Where(s => s.vol1 == goal || s.vol2 == goal))
            {
                found = s;
                isFound = true;
                break;
            }
        }
        return found.vol1 == goal 
            ? new TwoBucketResult { Moves = steps, GoalBucket = Bucket.One, OtherBucket = found.vol2 } 
            : new TwoBucketResult { Moves = steps, GoalBucket = Bucket.Two, OtherBucket = found.vol1 };
    }
    
    private static int Gcd(uint a, uint b)
    // Adapted from:
    // https://stackoverflow.com/questions/18541832/c-sharp-find-the-greatest-common-divisor
    {
        while (a != 0 && b != 0)
        {
            if (a > b)
                a %= b;
            else
                b %= a;
        }
        return (int)(a | b);
    }
}
